File: EUREKABANK_SOAP_JAVA_BDD/01 SERVIDOR/Eurobank_Soap_Java/src/java/ec/edu/monster/servicio/EurekaService.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.monster.servicio;

import ec.edu.monster.db.AccesoDB;
import ec.edu.monster.modelo.Movimiento;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
/**
 *
 * @author josue
 */
public class EurekaService {
    
        public List<Movimiento> leerMovimientos(String cuenta) {
        Connection cn = null;
        List<Movimiento> lista = new ArrayList<>();
        String sql = "SELECT \n"
            + " m.chr_cuencodigo cuenta, \n"
            + " m.int_movinumero nromov, \n"
            + " m.dtt_movifecha fecha, \n"
            + " t.vch_tipodescripcion tipo, \n"
            + " t.vch_tipoaccion accion, \n"
            + " m.dec_moviimporte importe \n"
            + "FROM tipomovimiento t INNER JOIN movimiento m \n"
            + "ON t.chr_tipocodigo = m.chr_tipocodigo \n"
            + "WHERE m.chr_cuencodigo = ? \n"
            + "ORDER BY m.int_movinumero DESC";

        try {
            cn = AccesoDB.getConnection();
            PreparedStatement pstm = cn.prepareStatement(sql);
            pstm.setString(1, cuenta);
            ResultSet rs = pstm.executeQuery();

            while (rs.next()) {
                Movimiento rec = new Movimiento();
                rec.setCuenta(rs.getString("cuenta"));
                rec.setNromov(rs.getInt("nromov"));
                rec.setFecha(rs.getDate("fecha"));
                rec.setTipo(rs.getString("tipo"));
                rec.setAccion(rs.getString("accion"));
                rec.setImporte(rs.getDouble("importe"));

                lista.add(rec);
            }
            rs.close();
        } catch (SQLException e) {
            throw new RuntimeException(e.getMessage());
        } finally {
            try {
                if (cn != null) cn.close();
            } catch (Exception e) {
            }
        }
        return lista;
    }
        
    public void registrarDeposito(String cuenta, double importe, String codEmp) {
        ejecutarMovimiento(cuenta, importe, codEmp, "003", true);
    }

    public void registrarRetiro(String cuenta, double importe, String codEmp) {
        ejecutarMovimientoDescuento(cuenta, importe, codEmp, "004", false);
    }    
    
    public void registrarTransferencia(String cuentaOrigen, String cuentaDestino, double importe, String codEmp) {
        Connection cn = null;
        try {
            cn = AccesoDB.getConnection();
            cn.setAutoCommit(false);

            // Retirar de la cuenta de origen
            ejecutarMovimientoInternoDescuento(cuentaOrigen, importe, codEmp, "009", cn, false);

            // Depositar en la cuenta de destino
            ejecutarMovimientoInterno(cuentaDestino, importe, codEmp, "008", cn, true);

            // Confirmar la transacción
            cn.commit();
        } catch (SQLException e) {
            try {
                if (cn != null) cn.rollback();
            } catch (Exception ex) {
            }
            throw new RuntimeException(e.getMessage());
        } finally {
            try {
                if (cn != null) cn.close();
            } catch (Exception e) {
            }
        }
    }
    
    private void ejecutarMovimiento(String cuenta, double importe, String codEmp, String tipoMov, boolean permiteNegativo) {
        Connection cn = null;
        try {
            cn = AccesoDB.getConnection();
            cn.setAutoCommit(false);
            ejecutarMovimientoInterno(cuenta, importe, codEmp, tipoMov, cn, permiteNegativo);
            cn.commit();
        } catch (SQLException e) {
            try {
                if (cn != null) cn.rollback();
            } catch (Exception ex) {
            }
            throw new RuntimeException(e.getMessage());
        } finally {
            try {
                if (cn != null) cn.close();
            } catch (Exception e) {
            }
        }
    }

    private void ejecutarMovimientoDescuento(String cuenta, double importe, String codEmp, String tipoMov, boolean permiteNegativo) {
        Connection cn = null;
        try {
            cn = AccesoDB.getConnection();
            cn.setAutoCommit(false);
            ejecutarMovimientoInternoDescuento(cuenta, importe, codEmp, tipoMov, cn, permiteNegativo);
            cn.commit();
        } catch (SQLException e) {
            try {
                if (cn != null) cn.rollback();
            } catch (Exception ex) {
            }
            throw new RuntimeException(e.getMessage());
        } finally {
            try {
                if (cn != null) cn.close();
            } catch (Exception e) {
            }
        }
    }

    private void ejecutarMovimientoInterno(String cuenta, double importe, String codEmp, String tipoMov, Connection cn, boolean permiteNegativo) throws SQLException {
        String sql = "SELECT dec_cuensaldo, int_cuencontmov "
                + "FROM cuenta "
                + "WHERE chr_cuencodigo = ? AND vch_cuenestado = 'ACTIVO' FOR UPDATE";
        PreparedStatement pstm = cn.prepareStatement(sql);
        pstm.setString(1, cuenta);
        ResultSet rs = pstm.executeQuery();

        if (!rs.next()) {
            throw new SQLException("ERROR: Cuenta no existe o no está activa.");
        }
        double saldo = rs.getDouble("dec_cuensaldo");
        int cont = rs.getInt("int_cuencontmov");
        rs.close();
        pstm.close();

        saldo += importe;
        if (!permiteNegativo && saldo < 0) {
            throw new SQLException("ERROR: Saldo insuficiente.");
        }

        cont++;
        sql = "UPDATE cuenta SET dec_cuensaldo = ?, int_cuencontmov = ? WHERE chr_cuencodigo = ?";
        pstm = cn.prepareStatement(sql);
        pstm.setDouble(1, saldo);
        pstm.setInt(2, cont);
        pstm.setString(3, cuenta);
        pstm.executeUpdate();
        pstm.close();

        sql = "INSERT INTO movimiento(chr_cuencodigo, int_movinumero, dtt_movifecha, chr_emplcodigo, chr_tipocodigo, dec_moviimporte) "
                + "VALUES (?, ?, SYSDATE(), ?, ?, ?)";
        pstm = cn.prepareStatement(sql);
        pstm.setString(1, cuenta);
        pstm.setInt(2, cont);
        pstm.setString(3, codEmp);
        pstm.setString(4, tipoMov);
        pstm.setDouble(5, importe);
        pstm.executeUpdate();
        pstm.close();
    }
    
    
    private void ejecutarMovimientoInternoDescuento(String cuenta, double importe, String codEmp, String tipoMov, Connection cn, boolean permiteNegativo) throws SQLException {
        String sql = "SELECT dec_cuensaldo, int_cuencontmov "
                + "FROM cuenta "
                + "WHERE chr_cuencodigo = ? AND vch_cuenestado = 'ACTIVO' FOR UPDATE";
        PreparedStatement pstm = cn.prepareStatement(sql);
        pstm.setString(1, cuenta);
        ResultSet rs = pstm.executeQuery();

        if (!rs.next()) {
            throw new SQLException("ERROR: Cuenta no existe o no está activa.");
        }
        double saldo = rs.getDouble("dec_cuensaldo");
        int cont = rs.getInt("int_cuencontmov");
        rs.close();
        pstm.close();

        saldo -= importe;
        if (!permiteNegativo && saldo < 0) {
            throw new SQLException("ERROR: Saldo insuficiente.");
        }

        cont++;
        sql = "UPDATE cuenta SET dec_cuensaldo = ?, int_cuencontmov = ? WHERE chr_cuencodigo = ?";
        pstm = cn.prepareStatement(sql);
        pstm.setDouble(1, saldo);
        pstm.setInt(2, cont);
        pstm.setString(3, cuenta);
        pstm.executeUpdate();
        pstm.close();

        sql = "INSERT INTO movimiento(chr_cuencodigo, int_movinumero, dtt_movifecha, chr_emplcodigo, chr_tipocodigo, dec_moviimporte) "
                + "VALUES (?, ?, SYSDATE(), ?, ?, ?)";
        pstm = cn.prepareStatement(sql);
        pstm.setString(1, cuenta);
        pstm.setInt(2, cont);
        pstm.setString(3, codEmp);
        pstm.setString(4, tipoMov);
        pstm.setDouble(5, importe);
        pstm.executeUpdate();
        pstm.close();
    }    
}

File: Deberes_EurekaBank_Arquitectura-entregables/EUREKABANK_RESTFULL_DOTNET_BDD_GR06/01 SERVIDOR/Eurekabank_Restfull_Dotnet/Eurekabank_Restfull_Dotnet/Services/Imp/EurekaService.cs
using Eurekabank_Restfull_Dotnet.Models;
using Eurekabank_Restfull_Dotnet.Services.Interfaces;
using System.Data.SqlClient;

namespace Eurekabank_Restfull_Dotnet.Services.Imp
{
    public class EurekaService : IEurekaService
    {
        private readonly IConfiguration _configuration;
        private string ConnectionString => _configuration.GetConnectionString("EurekaDB");

        public EurekaService(IConfiguration configuration)
        {
            _configuration = configuration;
        }

        public List<Movimiento> ListarPorCuenta(string cuenta)
        {
            var lista = new List<Movimiento>();
            string sql = @"
                SELECT
                    m.chr_cuencodigo cuenta,
                    m.int_movinumero nromov,
                    m.dtt_movifecha fecha,
                    t.vch_tipodescripcion tipo,
                    t.vch_tipoaccion accion,
                    m.dec_moviimporte importe
                FROM tipomovimiento t
                INNER JOIN movimiento m ON t.chr_tipocodigo = m.chr_tipocodigo
                WHERE m.chr_cuencodigo = @cuenta";

            using (SqlConnection cn = new SqlConnection(ConnectionString))
            {
                cn.Open();
                SqlCommand cmd = new SqlCommand(sql, cn);
                cmd.Parameters.AddWithValue("@cuenta", cuenta);
                SqlDataReader dr = cmd.ExecuteReader();

                while (dr.Read())
                {
                    lista.Add(new Movimiento
                    {
                        Cuenta = dr["cuenta"].ToString(),
                        NroMov = Convert.ToInt32(dr["nromov"]),
                        Fecha = Convert.ToDateTime(dr["fecha"]),
                        Tipo = dr["tipo"].ToString(),
                        Accion = dr["accion"].ToString(),
                        Importe = Convert.ToDouble(dr["importe"])
                    });
                }
            }
            return lista;
        }

        public void RegistrarDeposito(string cuenta, double importe, string codEmp = "0001")
        {
            EjecutarMovimiento(cuenta, importe, codEmp, "003", true);
        }

        public void RegistrarRetiro(string cuenta, double importe, string codEmp = "0001")
        {
            EjecutarMovimiento(cuenta, -importe, codEmp, "004", false);
        }

        public void RegistrarTransferencia(string cuentaOrigen, string cuentaDestino, double importe, string codEmp = "0001")
        {
            using (SqlConnection cn = new SqlConnection(ConnectionString))
            {
                cn.Open();
                SqlTransaction tx = cn.BeginTransaction();
                try
                {
                    EjecutarMovimientoInterno(cuentaOrigen, -importe, codEmp, "009", cn, tx, false);
                    EjecutarMovimientoInterno(cuentaDestino, importe, codEmp, "008", cn, tx, true);
                    tx.Commit();
                }
                catch
                {
                    tx.Rollback();
                    throw;
                }
            }
        }

        private void EjecutarMovimiento(string cuenta, double importe, string codEmp, string tipoMov, bool permiteNegativo)
        {
            using (SqlConnection cn = new SqlConnection(ConnectionString))
            {
                cn.Open();
                SqlTransaction tx = cn.BeginTransaction();
                try
                {
                    EjecutarMovimientoInterno(cuenta, importe, codEmp, tipoMov, cn, tx, permiteNegativo);
                    tx.Commit();
                }
                catch
                {
                    tx.Rollback();
                    throw;
                }
            }
        }

        private void EjecutarMovimientoInterno(string cuenta, double importe, string codEmp, string tipoMov,
                                              SqlConnection cn, SqlTransaction tx, bool permiteNegativo)
        {
            string sql = "SELECT dec_cuensaldo, int_cuencontmov FROM cuenta WHERE chr_cuencodigo = @cuenta AND vch_cuenestado = 'ACTIVO'";
            SqlCommand cmd = new SqlCommand(sql, cn, tx);
            cmd.Parameters.AddWithValue("@cuenta", cuenta);
            SqlDataReader dr = cmd.ExecuteReader();

            if (!dr.Read())
            {
                dr.Close();
                throw new Exception("ERROR: Cuenta no existe o no está activa.");
            }

            double saldo = Convert.ToDouble(dr["dec_cuensaldo"]);
            int cont = Convert.ToInt32(dr["int_cuencontmov"]);
            dr.Close();

            saldo += importe;
            if (!permiteNegativo && saldo < 0)
            {
                throw new Exception("ERROR: Saldo insuficiente.");
            }

            cont++;

            cmd = new SqlCommand("UPDATE cuenta SET dec_cuensaldo = @saldo, int_cuencontmov = @cont WHERE chr_cuencodigo = @cuenta", cn, tx);
            cmd.Parameters.AddWithValue("@saldo", saldo);
            cmd.Parameters.AddWithValue("@cont", cont);
            cmd.Parameters.AddWithValue("@cuenta", cuenta);
            cmd.ExecuteNonQuery();

            cmd = new SqlCommand(@"
                INSERT INTO movimiento(chr_cuencodigo, int_movinumero, dtt_movifecha, chr_emplcodigo, chr_tipocodigo, dec_moviimporte)
                VALUES (@cuenta, @nro, GETDATE(), @codEmp, @tipoMov, @importe)", cn, tx);
            cmd.Parameters.AddWithValue("@cuenta", cuenta);
            cmd.Parameters.AddWithValue("@nro", cont);
            cmd.Parameters.AddWithValue("@codEmp", codEmp);
            cmd.Parameters.AddWithValue("@tipoMov", tipoMov);
            cmd.Parameters.AddWithValue("@importe", Math.Abs(importe));
            cmd.ExecuteNonQuery();
        }
    }
}
